### ISSUE TITLE
BCAL reconstructed energy sometimes broken on startup: multicore only

### ISSUE THREAD

**Issue by jonzarling** [JeffersonLab/HDGeant4#143]
A problem was seen and reported for BCAL reconstructed photon energies in https://github.com/JeffersonLab/HDGeant4/issues/138. Unfortunately, I blamed the run number, which turns out to be a red herring. Best I can tell, the issue is actually random on job startup: either the energy response will be correct or it will be really bad.

To see this, I tried running several small photon gun jobs. I used 2 GeV photons fired at 60 degrees in theta. I generated 10,000 events of a particular run number, and repeated this for several jobs. Results of the reconstructed energy distribution are shown below for a few run numbers. You can see that a few topologies arise.

This issue is only observed in hdgeant4. I do not see the issue arise at all in geant3 MC.

![image](https://user-images.githubusercontent.com/17165552/73359734-821d9000-426f-11ea-9323-b78b910eeaa7.png)

![image](https://user-images.githubusercontent.com/17165552/73359754-8ba6f800-426f-11ea-9af9-575e0191e05b.png)

![image](https://user-images.githubusercontent.com/17165552/73359780-982b5080-426f-11ea-8ca5-9e5683b299c2.png)

**Comment by faustus123** [comment]
Hi Jon,

  Is each file a separate job? Curious that it is always the first 3 files that show an issue.
Regards,
-David

-------------------------------------------------------------
David Lawrence Ph.D.
Staff Scientist, Thomas Jefferson National Accelerator Facility
Newport News, VA
davidl@jlab.org<mailto:davidl@jlab.org>
(757) 269-5567 W
(757) 746-6697 C


On Jan 29, 2020, at 8:16 AM, Jonathan Zarling <notifications@github.com<mailto:notifications@github.com>> wrote:


A problem was seen and reported for BCAL reconstructed photon energies in #138<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_JeffersonLab_HDGeant4_issues_138&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=i0p-C_T1RieVsCZl8ZnUyQ&m=H5qkK4c150l77pEVI5gq975j2W8jSBl9KdVGBbXamfQ&s=9AP8BnpW4nCaUI0O6462HRjxl2FUy-zq3pelKn-8M5s&e=>. Unfortunately, I blamed the run number, which turns out to be a red herring. Best I can tell, the issue is actually random on job startup: either the energy response will be correct or it will be really bad.

To see this, I tried running several small photon gun jobs. I used 2 GeV photons fired at 60 degrees in theta. I generated 10,000 events of a particular run number, and repeated this for several jobs. Results of the reconstructed energy distribution are shown below for a few run numbers. You can see that a few topologies arise.

This issue is only observed in hdgeant4. I do not see the issue arise at all in geant3 MC.

[image]<https://urldefense.proofpoint.com/v2/url?u=https-3A__user-2Dimages.githubusercontent.com_17165552_73359734-2D821d9000-2D426f-2D11ea-2D9323-2Db78b910eeaa7.png&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=i0p-C_T1RieVsCZl8ZnUyQ&m=H5qkK4c150l77pEVI5gq975j2W8jSBl9KdVGBbXamfQ&s=KEEaXcTN9Uh3Hr7V9rmtSblNPJBf6JEn8OYIC9OT394&e=>

[image]<https://urldefense.proofpoint.com/v2/url?u=https-3A__user-2Dimages.githubusercontent.com_17165552_73359754-2D8ba6f800-2D426f-2D11ea-2D9af9-2D575e0191e05b.png&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=i0p-C_T1RieVsCZl8ZnUyQ&m=H5qkK4c150l77pEVI5gq975j2W8jSBl9KdVGBbXamfQ&s=K42B18ZMBdxmCqFXEBjkcvKKQCAJ8MIzRo_6SpLVJhw&e=>

[image]<https://urldefense.proofpoint.com/v2/url?u=https-3A__user-2Dimages.githubusercontent.com_17165552_73359780-2D982b5080-2D426f-2D11ea-2D8ca5-2D9e5683b299c2.png&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=i0p-C_T1RieVsCZl8ZnUyQ&m=H5qkK4c150l77pEVI5gq975j2W8jSBl9KdVGBbXamfQ&s=lI_wkDUAM-JK36WLj9D32-9fKPxTJntAa8YKGgY7b30&e=>

**Comment by jonzarling** [comment]
Hmmm looks like the files were getting sorted at some point, maybe they were getting fed in by file size order. Here's the same thing, except the filenames are given in terms of their actual job number. 

![image](https://user-images.githubusercontent.com/17165552/73361432-f279e080-4272-11ea-8fe4-1a3de8d7cb66.png)
![image](https://user-images.githubusercontent.com/17165552/73361454-f9a0ee80-4272-11ea-9e6f-dab70eaa0a93.png)
![image](https://user-images.githubusercontent.com/17165552/73361471-01f92980-4273-11ea-8d18-78abbffd405e.png)

**Comment by sdobbs** [comment]
Are you running in multithreaded mode?

On Wed, Jan 29, 2020 at 8:42 AM Jonathan Zarling <notifications@github.com<mailto:notifications@github.com>> wrote:

Hmmm looks like the files were getting sorted at some point, maybe they were getting fed in by file size order. Here's the same thing, except the filenames are given in terms of their actual job number.

[image]<https://user-images.githubusercontent.com/17165552/73361432-f279e080-4272-11ea-8fe4-1a3de8d7cb66.png>
[image]<https://user-images.githubusercontent.com/17165552/73361454-f9a0ee80-4272-11ea-9e6f-dab70eaa0a93.png>
[image]<https://user-images.githubusercontent.com/17165552/73361471-01f92980-4273-11ea-8d18-78abbffd405e.png>

**Comment by jonzarling** [comment]
Yep, looks like this is a multi-core issue. Here's the results of doing the same thing single threaded. Thanks for the catch Sean.

![image](https://user-images.githubusercontent.com/17165552/73554624-1592c480-441a-11ea-9c8c-763fada0ae26.png)
![image](https://user-images.githubusercontent.com/17165552/73554650-1f1c2c80-441a-11ea-9af2-8dcc570e5db9.png)
![image](https://user-images.githubusercontent.com/17165552/73554660-25aaa400-441a-11ea-82ff-6d83b7fa7aaa.png)

**Comment by sdobbs** [comment]
Good progress!  Could you please upload the configuration files you are using, so that Richard or someone else can try to reproduce the problem?

One other request:  do you still have the logs from the jobs that were giving the bad results?  It would be good to see if there were any errors in loading CCDB constants.  Maybe there were some problems with that, and it would be good to know if these were silent or not

**Comment by jonzarling** [comment]
Interesting that even the multi-core jobs that looked most ok had too much of a left tail. It appears the BCAL response changes either moderately or very severely multi-threaded, but I don't think it could ever be considered acceptable. I made these samples through MCWrapper which I assume should handle multi-core stuff properly on its end.

**Comment by jonzarling** [comment]
The logs should still exist, and I can give Richard my configuration as well without too much trouble. 

@rjones30 let me know what you'd like. Is a control.in enough, or would you like the MCWrapper config too?

**Comment by sdobbs** [comment]
My guess would be that there is some problem loading the CCDB constants
(which we have seen lately), and execution just continued.  So having some
logs would help with that.

**Comment by jonzarling** [comment]
Sure, here are some example logs from a particularly bad 4 core example above, along with a 1 core log for comparison. I guess github doesn't log text file uploads, let me know if docx format is ok or not.

[Run70000_0_1coreGOOD.docx](https://github.com/JeffersonLab/HDGeant4/files/4140457/Run70000_0_1coreGOOD.docx)


[Run30300_000_4cores_BAD.docx](https://github.com/JeffersonLab/HDGeant4/files/4140459/Run30300_000_4cores_BAD.docx)

**Comment by rjones30** [comment]
Hello Jon,

The control.in contains everything, but I also need the input MC file if
you are using one. Otherwise if it is just the internal generator,
control.in should be enough.

-Richard

**Comment by rjones30** [comment]
Hello Jon,

The control.in contains everything, but I also need the input MC file if
you are using one. Otherwise if it is just the internal generator,
control.in should be enough.

-Richard

**Comment by jonzarling** [comment]
Hi all,

Apologies for the delay. A family emergency came up about two weeks ago, I am just returning to this.

Here is an example control.in. Note that the TRAJECTORIES flag is on, though that shouldn't affect anything.


[control.in.txt](https://github.com/JeffersonLab/HDGeant4/files/4205152/control.in.txt)

**Comment by sdobbs** [comment]
Hi Jon,

I took another look at this problem, and could recreate it with the particle gun just fine, but didn't see much difference between single-threaded and multi-threaded mode when running a small amount of events generated with an external generator (J/psi).

Would you be able to try to do such a check as well?  Maybe your eta -> 3pi0 analysis sends enough photons into the BCAL (especially if you generate more events at large t), and you could compare the E(recon) - E(thrown) distribution for photons that were in the BCAL.

**Comment by rjones30** [comment]
I have made a deep dive into this issue, and found a number of issues related to thread-safety in our existing hdgeant4 codebase. Most of the problem lies within the G4 library itself, but fixing these also motivated some changes to the GlueX codebase for hdgeant4 to make it more robust against possible cross-thread interactions. The primary problems are listed here in a nutshell.

1. thread-shared fcopyNo member in G4PVDivision - this data member is overwritten during tracking to save the state of the navigation. It needed to be made both instance-local and thread-local. I used the existing implementation of similar functionality in G4PVReplica to implement a similar thread-local storage for fcopyNo in G4PVDivision.
2. internal state in G4MagHelicalStepper is not cleared between events, and carries over from one event to the next affected the output from G4HelixMixedStepper as it tracks charged particles in the magnetic field of the solenoid. For this I had to add a new public method ResetState() to be called from G4FieldManagerStore::ClearAllChordFinderState. This is called at the start of each track, which is where it makes sense to reset the helix parameters and start again from zero.

In addition to this work on the base G4 library code, I also made some fixes to our own hdgeant4 codebase.

1. previously the simulation would not write out events that turned out not to have any detector hits. Why this should ever happen with 2GeV photons at 60 degrees is a separate question (see issues above) but it is confusing to generate 10k events and see only 9235 events in the output file. This is fixed.
2. previously a single DMagneticFieldMap object was used for magnetic field lookups by all threads, one for the solenoid and one for the PS. This is done using the GetField method which is const, so I thought it should be re-entrant. But I found online some people warning that things that you might think are re-entrant (eg. find on a std::map) are not guaranteed to be so by the standard. So I wrote to David Lawrence asking if const members of JANA objects are supposed to be reentrant and he has not replied, so I guess this question does not have a clear and obvious answer. To remove this risk, I now require each thread to have its own private copy of each DMagneticFieldMap object, which costs extra memory but not too much. Of course, I never considered adding a mutex -- that would be a performance killer!!
3. added support for specifying a constant solenoid field in control.in, using BFIELDTYPE 'Const' <value>. Previously the 'Const' value for BFIELDTYPE was supported by hdgeant4, but the value could not be specified!!
4. other enhancements to help me debug stuff.

Results will be shown in follow-up entries, but here is a summary in plain words.

1. Everything has changed significantly in the bcal response to a 2GeV photon at 60 degrees. This includes both running with just one thread and with more than one thread. The tail to low energy has decreased.
2. There is a MUCH better agreement now with hdgeant. The difference in the mean is at the level of 0.5% without the fudge factor that currently exists in the hdgeant4 master branch. I am not going to change the fudge factor until I get agreement from Mark D. and Igal, this is just to point out that this was a symptom of the bug in G4. 
3. The results are now identical when you run starting from the same set of seeds, regardless of the number of threads you use. This means track-for-track identical, all except for the event numbers which always come out in increasing order and so can be assigned differently to events between different runs with multiple threads, if you run with the particle gun. Of course, running with an external generator ALWAYS gives identical output from the same starting seeds, regardless of the threads used -- now.
4. The resolution of reconstructed shower energy in the bcal from simulation (both hdgeant and hdgeant4, which now agree) at 60 degrees seems to be a factor 2 better than what is reported in the NIM paper for the average bcal resolution for 2GeV photons. This might indicate that there is a strong position dependence in the energy resolution, or the energy smearing in the bcal may need some work following this update to hdgeant4.

**Comment by rjones30** [comment]
Here are my new results with a photon gun, similar to Jon's original simulation at the start of this issue.
  10k events, 2 GeV photons, coming from the beam-target interaction region, uniformly distributed along the target, emitted at 60 degrees polar angle, uniform in 2pi azimuthal angle.
The histograms are:
1. hdgeant (green histogram)
2. hdgeant4, MT build running with 1 thread (black histogram)
3. hdgeant4, MT build running with 1 thread, 100k events rescaled /10 (blue points)
4. hdgeant4, MT build running with 4 threads, 100k events rescaled /10 (red points) -- different starting seeds from #3

This is the total energy (truthHit sum) deposited in the bcal by the simulation.
![image](https://user-images.githubusercontent.com/7832920/86390180-9c67f980-bc65-11ea-9720-490ce84cc3aa.png)

This is the same thing without the (apparently spurious) bcal fudge factor:
![image](https://user-images.githubusercontent.com/7832920/86390259-bdc8e580-bc65-11ea-9e09-6c35f99bf20d.png)

I am removing the fudge factor for all remaining comparison plots. The above plots are made prior to mcsmear + reconstruction. Here is the total reconstructed shower energy in the bcal after mcsmear + reconstruction, at the level of the rest file output.
![image](https://user-images.githubusercontent.com/7832920/86390389-f5d02880-bc65-11ea-8480-c17056f1329a.png)

And here is what you get if you take only the most energetic reconstructed shower in the bcal, and drop any split-off showers.
![image](https://user-images.githubusercontent.com/7832920/86390441-100a0680-bc66-11ea-9b05-b732f2f45710.png)

**Comment by rjones30** [comment]
Some numbers from the above plot: what is the size of the remaining mismatch in average bcal shower energy between g3 and g4?

- raw truth energy before smearing: hdgeant is higher than hdgeant4 by 0.60%
- reconstructed bcal total shower energy: hdgeant is higher than hdgeant4 by 0.52%
- reconstructed bcal biggest shower energy: hdgeant is higher than hdgeant4 by 0.48%

**Comment by rjones30** [comment]
I believe that this issue can be closed.
-Richard

**Comment by jonzarling** [comment]
Hi Richard,

Sorry for the delay here, just wanted to verify myself that the issue is fixed. Indeed it is. Here's the same study as I used when I opened the issue. Everything is clean and consistent now between different jobs, for 1 or 4 cores.

![image](https://user-images.githubusercontent.com/17165552/87448006-41b99080-c5c9-11ea-9a1f-77c13a06673a.png)
![image](https://user-images.githubusercontent.com/17165552/87448044-4f6f1600-c5c9-11ea-82b2-ace16920cbee.png)
